src/explorer/connectionTreeItem.ts

```diff
@@ -59,63 +66,85 @@ export default class ConnectionTreeItem extends vscode.TreeItem
+    }
 
-        const dataService = this._connectionController.getActiveConnection();
-        dataService.listDatabases((err: any, databases: string[]) => {
-          if (err) {
-            return reject(`Unable to list databases: ${err}`);
-          }
+    if (this._childrenCacheIsUpToDate) {
+      return Promise.resolve(Object.values(this._childrenCache));
+    }
 
-          this._childrenCacheIsUpToDate = true;
+    return new Promise((resolve, reject) => {
+      const dataService = this._connectionController.getActiveConnection();
+      dataService.listDatabases((err: any, databases: string[]) => {
+        if (err) {
+          return reject(`Unable to list databases: ${err}`);
+        }
 
-          if (databases) {
-            this._childrenCache = databases.map(
-              ({ name }: any) => new DatabaseTreeItem(name, dataService)
-            );
-          } else {
-            this._childrenCache = [];
-          }
+        this._childrenCacheIsUpToDate = true;
+
+        if (databases) {
+          const pastChildrenCache = this._childrenCache;
+          this._childrenCache = {};
+
+          databases.forEach(({ name }: any) => {
+            if (pastChildrenCache[name]) {
+              // We create a new element here instead of reusing the cached one
+              // in order to ensure the expanded state is set.
+              this._childrenCache[name] = new DatabaseTreeItem(
+                name,
+                dataService,
+                pastChildrenCache[name].isExpanded,
+                pastChildrenCache[name].getChildrenCache()
+              );
+            } else {
+              this._childrenCache[name] = new DatabaseTreeItem(
+                name,
+                dataService,
+                false, // Collapsed.
+                {} // No existing cache.
+              );
+            }
+          });
+        } else {
+          this._childrenCache = {};
+        }
 
-          return resolve(this._childrenCache);
-        });
+        return resolve(Object.values(this._childrenCache));
       });
-    }
-
-    // Here we either want to return loading or nothing.
-    return Promise.resolve([]);
+    });
   }
 
   onDidCollapse(): void {
     this.isExpanded = false;
     this._childrenCacheIsUpToDate = false;
   }
 
-  onDidExpand(): void {
+  onDidExpand(): Promise<boolean> {
     this._childrenCacheIsUpToDate = false;
     this.isExpanded = true;
+
+    // If we aren't the active connection, we reconnect.
+    if (this._connectionController.getActiveConnectionInstanceId() !== this._connectionInstanceId) {
+      return new Promise(resolve => {
+        this._connectionController.connectWithInstanceId(this._connectionInstanceId).then(
+          () => resolve(true),
+          err => {
+            this.isExpanded = false;
+            vscode.window.showErrorMessage(err);
+            resolve(false);
+          }
+        );
+      });
+    }
+
+    return Promise.resolve(true);
```

Comment:
nit: switch this to an early return by moving `return Promise.resolve(true);` to top

```javascript
if (this._connectionController.getActiveConnectionInstanceId() === this._connectionInstanceId) {
  return Promise.resolve(true);
}
```